README.md
# wannatalkcore

Wannatalk Official Flutter plugin.

## Getting Started

Edit the `pubspec.yaml` file in your app directory to define the Wannatalk SDK dependency:

```yaml
dependencies:
  wannatalkcore: ^0.0.25
```

## iOS Setup
1. Request **WTService-Info.plist** from [wannatalk.ai](https://wannatalk.ai) and drag it into your application. When prompted, select **"Copy items if needed"** and continue.
2. Add below keys in your application's Info.plist

    ```
   <key>NSPhotoLibraryUsageDescription</key>
   <string>To save in-chat photos and videos</string>
   <key>NSContactsUsageDescription</key>
   <string>To locate friends already on the Wannatalk network</string>
   <key>NSCameraUsageDescription</key>
   <string>To take photos and videos to send to friends in-chat</string>
   <key>NSLocationWhenInUseUsageDescription</key>
   <string>To display it in map</string>
   <key>NSMicrophoneUsageDescription</key>
   <string>To record live photos and movies</string>
   ```
3. Add this line to your application pod file

    `pod 'WTExternalSDK', :git =>'https://github.com/edzehoo/WannatalkAPI-iOS.git', :tag => '1.8.1'`

4. That's it! Run the app

## To strip simulator architecture framework for your app submission
1. Copy **trim.sh** file into your project folder.
2. Create Run Script Phase in Build Phases of your application target.
3. Paste `"$SRCROOT/trim.sh"` inside the body of Run Script Phase.
4. Enable `Run script only when installing` and `Show environment variables in build log`.
<img src="Screenshots/Screenshot1.png" width="500">

## Android Setup
1. Request **wannatalk-services.json** from [wannatalk.ai](https://wannatalk.ai) and drag it into your application `assets` directory
<img src="Screenshots/Screenshot2.png" width="350">
2. Enable `multiDexEnabled` in your application **build.gradle**
   
   ```
   android {
       defaultConfig {
           // Enabling multidex support.
           multiDexEnabled true
       }
       ...
   }
   ```
3. Set colorPrimary, colorPrimaryDark and colorAccent attributes in app's color.xml (src/main/res/values/colors.xml)
4. Add `maven { url 'https://jitpack.io' }` in your project build.gradle
   
   ```
   allprojects {
       repositories {
           ...
                      
           maven { url 'https://jitpack.io' }
           
           ...
       }
       
   }
   ```
5. That's it! Run the app


# Usage
```dart
import 'package:wannatalkcore/wannatalkcore.dart';
```


## To link Wannatalk account
```dart
Wannatalkcore.login(onCompletion: (WTResult result) {
  if (result.success) {
  
  }
});
```

## To link Wannatalk account with user credentials

```dart
Wannatalkcore.silentLogin("<user_identifier>", { displayname: "name", key1: "Value1", key2: "Value2"}, onCompletion: (WTResult result) {
  if (result.success) {
  }
});
```

### To check login status
```dart
bool loggedIn = await Wannatalkcore.isUserLoggedIn;
```

## To unlink Wannatalk account
```dart
Wannatalkcore.logout(onCompletion: (WTResult result) {
});
```

### Wannatalk events

This callback will be triggered to notify events:
1. kWTEventTypeLogin: This event will be triggered when user login into wannatalk account. `evenResponse.success` indicates whether login successful or not. If fails, check error information at `eventResponse.error`.
2. kWTEventTypeLogout: This event will be triggered when user loggout from wannatalk account. `evenResponse.success` indicates whether logout successful or not. If fails, check error information at `eventResponse.error`.
3. kWTEventTypeProduct: This event will be triggered when user click on product image in chat page. `evenResponse.result` contains productID and storeID of the respective product image message.

```dart
    
Wannatalkcore.setMethodCallHandler(onReceivedEvent:(WTEventResponse eventResponse) {

    switch (eventResponse.eventType) {
        case WTEventResponse.kWTEventTypeProduct: {
            var result = eventResponse.result;
            if (result.success) {
                if (result.productID != null && result.storeID != null) {
                    print("ProductID: " + result.productID!);
                    print("StoreID: " + result.storeID!);
                }
            }
            else {
                // Error
                print(result.error);
            }
            break;
        }
    }
});
```


## To send text message to user
```dart

    String receiverUserIdentifier = "+60199299399499";  // Receiver mobile number with calling code
    
    String message = "Hi"; // Message you want to send to user
    Wannatalkcore.sendTextMessage(receiverUserIdentifier, message, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });

```



## To send product image to user
```dart

    String receiverUserIdentifier = "+60199299399499"; // Receiver mobile number with calling code

    String productID = "<productID>"; // Product identifier 
    String productName = "<productName>"; // Optional // Product name
    String productPrice = "<productPrice>"; // Optional // Product price
    String productImageUrl = "https://upload.wikimedia.org/wikipedia/commons/a/ab/Apple-logo.png"; // Product image url
    String caption = "Can share details about this product?"; // caption will be sent to receiver along with image
    String storeID = "<storeID>"; // Store/Shop identifier 

    WTChatInput chatInput = WTChatInput.Product(productID, productName,
            productPrice, productImageUrl,
            caption, storeID);
        
    Wannatalkcore.sendProductImage(receiverUserIdentifier, chatInput, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
    

```


## To contact organization support
```dart

    String orgID = "your_organization_id";
    String channelID = "your_chat_channel_id";
    String message = "Hi"; // This message will be sent upon opening chat
    String ticketName = "alpha"; // Ticket name
    bool closeOldTickets = false; // If true, closes all active tickets, else opens existing active ticket

    Wannatalkcore.contactOrganization(orgID, channelID, ticketName, closeOldTickets, message, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });

```



    
## HelpDesk
### To load your organization profile
```dart
Wannatalkcore.loadOrganizationProfile(autoOpenChat, onCompletion: (WTResult result){
});
```

## Collaboration
### To view all chats
```dart
Wannatalkcore.loadChats(onCompletion: (WTResult result) {
});
```
    
### To view all users
```dart
Wannatalkcore.loadUsers(onCompletion: (WTResult result) {
});
```

## Other

### To show or hide guide button
```dart
WannatalkConfig.showGuideButton(true); // default = true
```
### To enable or disable sending audio message
```dart
WannatalkConfig.allowSendAudioMessage(false);  // default = true
```
### To show or hide add participants option in new ticket page and chat item profile page
```dart
WannatalkConfig.allowAddParticipants(false);    // default = true
```
### To show or hide remove participants option in chat item profile
```dart
WannatalkConfig.allowRemoveParticipants(false); // default = false
```
### To show or hide welcome message
```dart
WannatalkConfig.showWelcomeMessage(false);            // default = false
```
### To show or hide Profile Info page
```dart
WannatalkConfig.showProfileInfoPage(false);           // default = true
```
### To create auto tickets: 
Chat ticket will create automatically when auto tickets is enabled, otherwise default ticket creation page will popup
```dart
WannatalkConfig.enableAutoTickets(true);           // default = false
```
### To show or hide close chat button in chat page
```dart
WannatalkConfig.showExitButton(true);               // default = false
```
### To show or hide participants in chat profile page
```dart
WannatalkConfig.showChatParticipants(false);          // default = true
```
### To enable or disbale chat profile page
```dart
WannatalkConfig.enableChatProfile(false);           // default = true
```
### To allow modify in chat profile page
```dart
WannatalkConfig.allowModifyChatProfile(false);       // default = true
```
### To set Inactive chat timeout:  
Chat session will end if user is inactive for timeout interval duration. If timeout interval is 0, chat session will not end automatically. The default timout interval is 1800 seconds (30 minutes).
```dart
double timeoutInterval = 1800; // Default Value: 1800 seconds ~ 30 minutes
WannatalkConfig.setInactiveChatTimeoutInterval(timeoutInterval);
```


CHANGELOG.md
## 0.0.1

- initial release.

## 0.0.2

- Fixes

## 0.0.3

- Migrated to AndroidX
- Bug fixes

## 0.0.4

- Migrate to null safety

## 0.0.5

- Migrate to null safety

## 0.0.6

- Fixes

## 0.0.7

- Fixes

## 0.0.8

- Fixes

## 0.0.9

- Fixes

## 0.0.10

- Fixes

## 0.0.12

- Fixes

## 0.0.13

- Customize fonts
- Customize chat bubble colors

## 0.0.14

- Customize fonts
- Customize chat bubble colors

## 0.0.15

- Customize fonts
- Customize chat bubble colors
- Bug fixes

## 0.0.16

- Customize fonts
- Customize chat bubble colors
- Bug fixes

## 0.0.17

- Customize fonts
- Customize chat bubble colors
- Bug fixes

## 0.0.18

- Image upload crash fix

## 0.0.19

- Hide phone number in topics list

## 0.0.20

- Fix: Display server name in topics list

## 0.0.21

- Fix: Contacts reload issue

## 0.0.22

- API path and Hub path are required fields in config file


## 0.0.23

- Fixes

## 0.0.24

- allow open org chat

## 0.0.25

- allow open org chat


example/lib/main.dart
import 'dart:ffi';

import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:wannatalkcore/wannatalkcore.dart';
import 'package:flutter/foundation.dart';
import 'dart:developer' as developer;
import 'package:flutter/material.dart';
import 'dart:io' show Platform;

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool _userLoggedIn = false;

  @override
  void initState() {
    super.initState();
    initPlatformState();

    Wannatalkcore.setMethodCallHandler(onReceivedEvent:(WTEventResponse eventResponse) {

      switch (eventResponse.eventType) {
        case WTEventResponse.kWTEventTypeProduct: {
          var result = eventResponse.result;

          if (result.success) {
            if (result.productID != null && result.storeID != null) {
              print("ProductID: " + result.productID!);
              print("StoreID: " + result.storeID!);
            }
          }
          else {
            // Error
            print(result.error);
          }



          break;
        }
        case WTEventResponse.kWTEventTypeLoadStorePage: {
          var result = eventResponse.result;

          if (result.success) {
            if (result.userIdentifier != null) {
              print("User PhoneNumber: " + result.userIdentifier!);
            }
          }
          else {
            // Error
            print(result.error);
          }
          break;
        }

      }

    });
  }

  // Platform messages are asynchronous, so we intialize in an async method.
  Future<void> initPlatformState() async {
    bool loggedIn = await Wannatalkcore.isUserLoggedIn;

    if (!mounted) return;

    WannatalkConfig.showGuideButton(false);
    WannatalkConfig.showProfileInfoPage(false);
    WannatalkConfig.allowSendAudioMessage(false);
    WannatalkConfig.allowAddParticipants(false);
    WannatalkConfig.enableAutoTickets(true);

    WannatalkConfig.showExitButton(true);
    WannatalkConfig.showHomeButton(true);
    WannatalkConfig.enableChatProfile(false);
    WannatalkConfig.setChatHeaderColor("#1475a5");
    WannatalkConfig.setChatTitleColor("#FFFFFF");
    WannatalkConfig.setChatBGColor("#FFFFFF");



    if (Platform.isAndroid) {
      // Android-specific code
      var fontNames = WTFontNames.init();
      fontNames.regular = "montserrat_regular.ttf";
      fontNames.bold = "montserrat_bold.ttf";
      fontNames.medium = "montserrat_medium.ttf";
      fontNames.italic = "montserrat_light.ttf";
      fontNames.light = "montserrat_light.ttf";

      WannatalkConfig.setFontNames(fontNames);



    } else if (Platform.isIOS) {
      // iOS-specific code
      var fontNames = WTFontNames.init();
      fontNames.regular = "Montserrat-Regular";
      fontNames.bold = "Montserrat-Bold";
      fontNames.medium = "Montserrat-Medium";
      fontNames.italic = "Montserrat-Light";
      fontNames.light = "Montserrat-Light";

      WannatalkConfig.setFontNames(fontNames);


    }

    setState(() {
      _userLoggedIn = loggedIn;
    });
  }

  void login(){
    Wannatalkcore.login(onCompletion: (WTResult result) {
      if (result.success) {
        setState(() {
          _userLoggedIn = true;
        });
      }

    });
  }

  void silentLogin() {
    String userIdentifier = "+919000220455";
    Map userInfo = { "displayname": "SG"};

    Wannatalkcore.silentLogin(userIdentifier, userInfo, onCompletion: (WTResult result) {
      if (result.success) {
        setState(() {
          _userLoggedIn = true;
        });
      }
    });
  }


  void logout() {
    Wannatalkcore.logout(onCompletion: (WTResult result) {
      if (result.success) {
        setState(() {
          _userLoggedIn = false;
        });
      }
    });
  }


  void presentOrgProfile(bool autoOpenChat) {
    Wannatalkcore.loadOrganizationProfile(autoOpenChat, 0, onCompletion: (WTResult result){
      if (result.success) {

      }
    });
  }

  void presentChats() {
    Wannatalkcore.loadChats(onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
  }

  void presentUsers() {
    Wannatalkcore.loadUsers(onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
  }


  void sendTextMessage() {

    String receiverUserIdentifier = "+60199299399499";  // Receiver mobile number with calling code

    String message = "Hi"; // Message you want to send to user
    Wannatalkcore.sendTextMessage(receiverUserIdentifier, message, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
  }

  void sendProductImage() {
    String receiverUserIdentifier = "+60199299399499"; // Receiver mobile number with calling code

    String productID = "<productID>"; // Product identifier
    String productName = "<productName>"; // Optional // Product name
    String productPrice = "<productPrice>"; // Optional // Product price
    String productImageUrl = "https://upload.wikimedia.org/wikipedia/commons/a/ab/Apple-logo.png"; // Product image url
    String caption = "Can share details about this product?"; // caption will be sent to receiver along with image
    String storeID = "<storeID>"; // Store/Shop identifier

    WTChatInput chatInput = WTChatInput.Product(productID, productName,
        productPrice, productImageUrl,
        caption, storeID);

    Wannatalkcore.sendProductImage(receiverUserIdentifier, chatInput, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
  }

  void contactOrganization() {

    String orgID = "360";
    String channelID = "399";
    String message = "Hi"; // This message will be sent upon opening chat
    String ticketName = "alpha"; // Ticket name
    bool closeOldTickets = false; // If true, closes all active tickets, else opens existing active ticket

    Wannatalkcore.contactOrganization(orgID, channelID, ticketName, closeOldTickets, message, onCompletion: (WTResult result) {
      if (result.success) {

      }
    });
  }

  @override
  Widget build(BuildContext context) {

    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Wannatalk Demo app'),
        ),
        body: new Center(
          child: new Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              if (!_userLoggedIn) new RaisedButton(
                onPressed: () {
                  login();
                },
                child: new Text('Login'),
              ),
              if (!_userLoggedIn) new RaisedButton(
                onPressed: () {
                  silentLogin();
                },
                child: new Text('Silent Login'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {logout();},
                child: new Text('Logout'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {presentOrgProfile(true);},
                child: new Text('Org Profile'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {presentChats();},
                child: new Text('Chats'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {presentUsers();},
                child: new Text('Users'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {sendTextMessage();},
                child: new Text('Send Text Message'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {sendProductImage();},
                child: new Text('Send Product Image'),
              ),
              if (_userLoggedIn) new RaisedButton(
                onPressed: () {contactOrganization();},
                child: new Text('Contact Organization'),
              )
            ],
          ),
        )
      ),
    );
  }

}


ios/Classes/WTConfigHandler.h
//
//  WTConfigHandler.h
//  Wannatalkflutter
//
//  Created by Srikanth Ganji on 27/03/20.
//

#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface WTConfigHandler : NSObject

+ (void) handleConfigMethodType:(NSInteger) methodType arguments:(NSDictionary *) args;


+ (void)ClearTempFiles;

// To show or hide guide button
+ (void)ShowGuideButton:(BOOL) show;               // default = YES

// To enable or disable sending audio message
+ (void)AllowSendAudioMessage:(BOOL) allow;  // default = YES

// To show or hide add participants option in new ticket page and chat item profile page
+ (void)AllowAddParticipants:(BOOL) allow;    // default = YES

// To show or hide remove participants option in chat item profile
+ (void)AllowRemoveParticipants:(BOOL) allow; // default = NO

// To show or hide welcome message
+ (void)ShowWelcomeMessage:(BOOL) show;            // default = NO

// To show or hide Profile Info page
+ (void)ShowProfileInfoPage:(BOOL) show;           // default = YES

// To create auto tickets:
//Chat ticket will create automatically when auto tickets is enabled, otherwise default ticket creation page will popup
+ (void)EnableAutoTickets:(BOOL) enable;           // default = NO

// To show or hide close chat button in chat page
+ (void)ShowExitButton:(BOOL) show;                // default = NO

// To show or hide participants in chat profile page
+ (void)ShowChatParticipants:(BOOL) show;          // default = YES

// To enable or disbale chat profile page
+ (void)EnableChatProfile:(BOOL) enable;           // default = YES

// To allow modify  in chat profile page
+ (void)AllowModifyChatProfile:(BOOL) allow;       // default = YES

// To set Inactive chat timeout:
//Chat session will end if user is inactive for timeout interval duration. If timeout interval is 0, chat session will not end automatically. The default timout interval is 1800 seconds (30 minutes).
+ (void)SetInactiveChatTimeoutInterval:(double) timeoutInterval;   // default = 1800 seconds (30 minutes).



+ (void) ShowHomeButton:(BOOL) value;
+ (void) SetChatHeaderColor:(NSString *) value;
+ (void) SetChatTitleColor:(NSString *) value;
+ (void) SetChatBGColor:(NSString *) value;
+ (void) SetFontNames:(NSDictionary *) value;

@end

NS_ASSUME_NONNULL_END


lib/wannatalkcore.dart
import 'dart:async';

import 'package:flutter/services.dart';


class WTFontNames {

  String? regular;
  String? bold;
  String? light;
  String? medium;
  String? italic;

  WTFontNames.init() {

  }
}

class WTEventResponse {

  static const  String cWTEvent= "wt_event";
  static const  int kWTEventTypeLogin= 2001;
  static const  int kWTEventTypeLogout= 2002;
  // static const  int kWTEventTypeOrder= 2003;
  static const  int kWTEventTypeProduct= 2004;
  static const  int kWTEventTypeLoadStorePage= 2005;


  late int eventType;
  late WTResult result;

  WTEventResponse.init() {
    this.eventType = -1;
    this.result = new WTResult.init();
  }
}

class WTResult {
  late bool success;
  String? error;
  String? productID;
  String? storeID;
  String? sellerOrderID;
  String? buyerOrderID;
  String? userIdentifier;

  WTResult(this.success, this.error);

  WTResult.init() {
    this.success = false;
    this.error = null;
  }

}


class WTChatInput {

  String source = "";
  String? imagePath;
  String? caption;
  String? productID;
  String? productName;
  String? productPrice;
  String? storeID;
  String? orderBuyerRef;
  String? orderSellerRef;

  WTChatInput.init(String source) {
    this.source = source;
  }

  static WTChatInput Product(String productID, String productName,
      String productPrice, String imagePath,
      String caption, String storeID) {

    WTChatInput chatInput = WTChatInput.init("Product");
    chatInput.productID = productID;
    chatInput.productName = productName;
    chatInput.productPrice = productPrice;
    chatInput.imagePath = imagePath;
    chatInput.caption = caption;
    chatInput.storeID = storeID;
    return chatInput;
  }
}


class Wannatalkcore {
  static const MethodChannel _channel =
      const MethodChannel('wannatalkcore');


  static late Function(WTEventResponse eventResponse) _onReceivedEvent;
  // , {required Function(WTResult result) onCompletion}
  static void setMethodCallHandler({required Function(WTEventResponse eventResponse) onReceivedEvent}) {
    _onReceivedEvent = onReceivedEvent;
    _channel.setMethodCallHandler(_wtEvent);
  }

  static Future<void> _wtEvent(MethodCall call) async {
    WTEventResponse eventResponse = new WTEventResponse.init();

    try {

      int eventType = call.arguments["eventType"];
      bool success = call.arguments["success"] == 1 || call.arguments["success"] == true;
      String? error = call.arguments["error"];

      eventResponse.eventType = eventType;
      eventResponse.result.success = success;
      eventResponse.result.error = error;

      eventResponse.result.productID = call.arguments[_cProductID];
      eventResponse.result.storeID = call.arguments[_cStoreID];
      eventResponse.result.sellerOrderID = call.arguments[_cSellerOrderID];
      eventResponse.result.buyerOrderID = call.arguments[_cBuyerOrderID];
      eventResponse.result.userIdentifier = call.arguments[_cWTUserIdentifier];

    } on Exception catch (e) {
      print("Failed to login: '${e.toString()}'.");

    }
    finally {
      _onReceivedEvent(eventResponse);
    }
  }

  static const  String _cWTDefaultMethod= "wannatalkDefaultMethod";
  static const  String _cWTUpdateConfig= "updateConfig";
  static const  String _cWTIsUserLoggedIn= "isUserLoggedIn";


  static const  String _cWTMethodType= "methodType";
  static const  String _cWTArgs= "args";
  static const  String _cWTUsername= "username";
  static const  String _cWTLocalImagePath= "localImagePath";

  static const  String _cProductID= "productID";
  static const  String _cStoreID= "storeID";
  static const  String _cSellerOrderID= "sellerOrderID";
  static const  String _cBuyerOrderID= "buyerOrderID";



  static const  String _cWTAutoOpenChat= "autoOpenChat";
  static const  String _cWTAutoOpenChannelId= "autoOpenChannelId";
  static const  String _cWTUserIdentifier= "userIdentifier";
  static const  String _cWTUserInfo= "userInfo";


  static const  int _kWTLoginMethod= 1001;
  static const  int _kWTSilentLoginMethod= 1002;
  static const  int _kWTLogoutMethod= 1003;
  static const  int _kWTOrgProfileMethod= 1004;
  static const  int _kWTChatListMethod= 1005;
  static const  int _kWTUserListMethod= 1006;
  static const  int _kWTUpdateUserNameMethod= 1007;
  static const  int _kWTUpdateUserImageMethod= 1008;
  static const  int _kWTIsUserLoggedIn= 1009;
  static const  int _kWTLoadUserChat= 1010;
  static const  int _kWTLoadUserChatWithImage= 1011;
  static const  int _kWTContactOrg= 1012;


  static final  String _cISource= "Source";
  static final  String _cIImage= "Image";
  static final  String _cICaption= "Caption";
  static final  String _cIProductID= "ProductID";
  static final  String _cIProductName= "ProductName";
  static final  String _cIProductPrice= "ProductPrice";
  static final  String _cIStoreID= "StoreID";

  static final  String _cIOrgID= "orgID";
  static final  String _cIChannelID= "channelID";
  static final  String _cITicketName= "ticketName";
  static final  String _cICloseOldTickets= "closeOldTickets";
  static final  String _cIMessage= "message";

  /// To check login status
  // static Future<bool> get isUserLoggedIn => _channel.invokeMethod(_cWTIsUserLoggedIn);
  static Future<bool> get isUserLoggedIn async {
    return await _channel.invokeMethod(_cWTIsUserLoggedIn);
  }
    
    
  /// To login into wannatalk account
  static Future<void> login({required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTLoginMethod, null, onCompletion: onCompletion);
  }

  /// To login into wannatalk account with user credentials
  static Future<void> silentLogin(String userIdentifier, Map userInfo, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTSilentLoginMethod, <String, dynamic>{_cWTUserIdentifier: userIdentifier, _cWTUserInfo: userInfo}, onCompletion: onCompletion);
  }

  /// To unlink Wannatalk account
  static Future<void> logout({required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTLogoutMethod, null, onCompletion: onCompletion);
  }
  /// To load your organization profile
  static Future<void> loadOrganizationProfile(bool autoOpenChat, int autoOpenChanenlId, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTOrgProfileMethod, <String, dynamic>{_cWTAutoOpenChat: autoOpenChat, _cWTAutoOpenChannelId: autoOpenChanenlId}, onCompletion: onCompletion);
  }
  /// To view all chats
  static Future<void> loadChats({required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTChatListMethod, null, onCompletion: onCompletion);
  }
  /// To view all users
  static Future<void> loadUsers({required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTUserListMethod, null, onCompletion: onCompletion);
  }

  /// To update user profile name
  static Future<void> updateUserName(String username, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTUpdateUserNameMethod, <String, dynamic>{_cWTUsername: username}, onCompletion: onCompletion);
  }

  /// To update user profile image
  static Future<void> updateUserImage(String localImagePath, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTUpdateUserImageMethod, <String, dynamic>{_cWTLocalImagePath: localImagePath}, onCompletion: onCompletion);
  }

//  static Future<String> _sendWannatalkMethod2(int kWTMethod, Object object, {Function(bool success, String error) onCompletion}) async {
//    String error = await _channel.invokeMethod(_cWTDefaultMethod, <String, dynamic>{_cWTMethodType: kWTMethod, _cWTArgs: object});
//    onCompletion(error == null, error);
//    return error;
//  }

  ///
  static Future<WTResult> _sendWannatalkMethod(int kWTMethod, Object? object, {required Function(WTResult result) onCompletion}) async {
    String? error = await _channel.invokeMethod(_cWTDefaultMethod, <String, dynamic>{_cWTMethodType: kWTMethod, _cWTArgs: object});
    WTResult result = new WTResult(error == null, error);
    onCompletion(result);
    return result;
  }

  ///
  static Future<void> _updateConfig(int kWTMethod, Object? object) async {
    _channel.invokeMethod(_cWTUpdateConfig, <String, dynamic>{_cWTMethodType: kWTMethod, _cWTArgs: object});
  }

  /// To send text message to user
  static Future<void> sendTextMessage(String receiverUserIdentifier, String message, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTLoadUserChat, <String, dynamic>{
      _cWTUserIdentifier: receiverUserIdentifier,
      _cIMessage: message
    }, onCompletion: onCompletion);
  }

  /// To send image message to user
  static Future<void> sendProductImage(String receiverUserIdentifier, WTChatInput chatInput, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTLoadUserChatWithImage, <String, dynamic>{
      _cWTUserIdentifier: receiverUserIdentifier,
      _cISource: chatInput.source,
      _cICaption: chatInput.caption,
      _cIImage: chatInput.imagePath,
      _cIProductID: chatInput.productID,
      _cIProductName: chatInput.productName,
      _cIProductPrice: chatInput.productPrice,
      _cIStoreID: chatInput.storeID
    }, onCompletion: onCompletion);
  }

  /// To contact organization
  static Future<void> contactOrganization(String orgID, String channelID, String ticketName, bool closeOldTickets, String message, {required Function(WTResult result) onCompletion}) async {
    _sendWannatalkMethod(_kWTContactOrg, <String, dynamic>{
      _cIOrgID: orgID,
      _cIChannelID: channelID,
      _cITicketName: ticketName,
      _cICloseOldTickets: closeOldTickets,
      _cIMessage: message
    }, onCompletion: onCompletion);
  }

}

/// Configurables
class WannatalkConfig {

  static const  int _kWTConfigClearTempFolder = 9001;
  static const  int _kWTConfigSetInactiveChatTimeout = 9002;
  static const  int _kWTConfigShowGuideButton = 9003;
  static const  int _kWTConfigAllowSendAudioMessage = 9004;
  static const  int _kWTConfigAllowAddParticipant = 9005;
  static const  int _kWTConfigAllowRemoveParticipants = 9006;
  static const  int _kWTConfigShowWelcomePage = 9007;
  static const  int _kWTConfigShowProfileInfoPage = 9008;
  static const  int _kWTConfigEnableAutoTickets = 9009;
  static const  int _kWTConfigShowExitButton = 9010;
  static const  int _kWTConfigShowChatParticipants = 9011;
  static const  int _kWTConfigEnableChatProfile = 9012;
  static const  int _kWTConfigAllowModifyChatProfile = 9013;
  static const  int _kWTConfigSetAgentQueueInterval = 9014;

  static const  int _kWTConfigShowHomeButton= 9015;
  static const  int _kWTConfigSetChatHeaderColor= 9016;
  static const  int _kWTConfigSetChatTitleColor= 9017;
  static const  int _kWTConfigSetChatBGColor= 9018;
  static const  int _kWTConfigSetFontNames= 9019;


  static const  String _cEnable= "enable";
  static const  String _cShow= "show";
  static const  String _cAllow= "allow";
  static const  String _cValue= "value";
  static const  String _cTimeoutInterval = "timeoutInterval";
  static const  String _cTimeInterval = "timeInterval";

  /// To clear temp files
  static Future<void> clearTempFiles(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigClearTempFolder, null);
  }

  /// To set Inactive chat timeout:
  ///
  /// Chat session will end if user is inactive for timeout interval duration.
  /// If timeout interval is 0, chat session will not end automatically.
  /// The default timout interval is 1800 seconds (30 minutes).
  static Future<void> setInactiveChatTimeout(double interval) async {
    Wannatalkcore._updateConfig(_kWTConfigSetInactiveChatTimeout, <String, dynamic>{_cTimeoutInterval: interval});
  }

  /// To set interval to check available agent  // default = 20 seconds
  static Future<void> setAgentQueueInterval(double interval) async {
    Wannatalkcore._updateConfig(_kWTConfigSetAgentQueueInterval, <String, dynamic>{_cTimeInterval: interval});
  }

  /// To show or hide guide button
  static Future<void> showGuideButton(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigShowGuideButton, <String, dynamic>{_cShow: show});
  }
  /// To enable or disable sending audio message
  static Future<void> allowSendAudioMessage(bool allow) async {
    Wannatalkcore._updateConfig(_kWTConfigAllowSendAudioMessage, <String, dynamic>{_cAllow: allow});
  }

  /// To show or hide add participants option in new ticket page and chat item profile page
  static Future<void> allowAddParticipants(bool allow) async {
    Wannatalkcore._updateConfig(_kWTConfigAllowAddParticipant, <String, dynamic>{_cAllow: allow});
  }

  /// To show or hide remove participants option in chat item profile
  static Future<void> allowRemoveParticipants(bool allow) async {
    Wannatalkcore._updateConfig(_kWTConfigAllowRemoveParticipants, <String, dynamic>{_cAllow: allow});
  }

  /// To show or hide welcome message
  static Future<void> showWelcomeMessage(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigShowWelcomePage, <String, dynamic>{_cShow: show});
  }

  /// To show or hide Profile Info page
  static Future<void> showProfileInfoPage(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigShowProfileInfoPage, <String, dynamic>{_cShow: show});
  }

  /// To create auto tickets:
  /// Chat ticket will create automatically when auto tickets is enabled,
  /// otherwise default ticket creation page will popup
  static Future<void> enableAutoTickets(bool enable) async {
    Wannatalkcore._updateConfig(_kWTConfigEnableAutoTickets, <String, dynamic>{_cEnable: enable});
  }

  /// To show or hide chat exit button in chat page
  static Future<void> showExitButton(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigShowExitButton, <String, dynamic>{_cShow: show});
  }

  /// To show or hide participants in chat profile page
  static Future<void> showChatParticipants(bool show) async {
    Wannatalkcore._updateConfig(_kWTConfigShowChatParticipants, <String, dynamic>{_cShow: show});
  }

  /// To enable or disbale chat profile page
  static Future<void> enableChatProfile(bool enable) async {
    Wannatalkcore._updateConfig(_kWTConfigEnableChatProfile, <String, dynamic>{_cEnable: enable});
  }

  /// To allow modify in chat profile page
  static Future<void> allowModifyChatProfile(bool allow) async {
    Wannatalkcore._updateConfig(_kWTConfigAllowModifyChatProfile, <String, dynamic>{_cAllow: allow});
  }


  static Future<void> showHomeButton(bool value) async {
    Wannatalkcore._updateConfig(_kWTConfigShowHomeButton, <String, dynamic>{_cValue: value});
  }
  static Future<void> setChatHeaderColor(String value) async {
    Wannatalkcore._updateConfig(_kWTConfigSetChatHeaderColor, <String, dynamic>{_cValue: value});
  }
  static Future<void> setChatTitleColor(String value) async {
    Wannatalkcore._updateConfig(_kWTConfigSetChatTitleColor, <String, dynamic>{_cValue: value});
  }
  static Future<void> setChatBGColor(String value) async {
    Wannatalkcore._updateConfig(_kWTConfigSetChatBGColor, <String, dynamic>{_cAllow: value});
  }
  static Future<void> setFontNames(WTFontNames fontNames) async {
    Wannatalkcore._updateConfig(_kWTConfigSetFontNames, <String, dynamic?>{
      "regular": fontNames.regular,
      "bold": fontNames.bold,
      "italic": fontNames.italic,
      "medium": fontNames.medium,
      "italic": fontNames.italic,
    });
  }

}
